components/selection/selection-mixin.js

```diff
@@ -0,0 +1,119 @@
+		this.addEventListener('d2l-selection-subscriber-unsubscribe', this._handleSelectionSubscriberUnsubscribe);
+		this.addEventListener('d2l-selection-checkbox-subscribe', this._handleSelectionCheckboxSubscribe);
+		this.addEventListener('d2l-selection-checkbox-unsubscribe', this._handleSelectionCheckboxUnsubscribe);
+	}
+
+	disconnectedCallback() {
+		super.disconnectedCallback();
+		this.removeEventListener('d2l-selection-change', this._handleSelectionChange);
+		this.removeEventListener('d2l-selection-select-all-change', this._handleSelectionSelectAllChange);
+		this.removeEventListener('d2l-selection-subscriber-subscribe', this._handleSelectionSubscriberSubscribe);
+		this.removeEventListener('d2l-selection-subscriber-unsubscribe', this._handleSelectionSubscriberUnsubscribe);
+		this.removeEventListener('d2l-selection-checkbox-subscribe', this._handleSelectionCheckboxSubscribe);
+		this.removeEventListener('d2l-selection-checkbox-unsubscribe', this._handleSelectionCheckboxUnsubscribe);
+	}
+
+	getSelectionInfo() {
+		const keys = [];
+		this._selectionSelectables.forEach(selectable => {
+			if (selectable.checked) keys.push(selectable.key);
+		});
+
+		let state = selectionStates.none;
+		if (keys.length > 0) {
+			if (keys.length === this._selectionSelectables.size) state = selectionStates.all;
+			else state = selectionStates.some;
+		}
+
+		return {
+			keys: keys,
+			state: state
+		};
+	}
+
+	_handleSelectionChange() {
+		this._updateSelectionSubscribers();
+	}
+
+	_handleSelectionCheckboxSubscribe(e) {
+		e.stopPropagation();
+		const target = e.composedPath()[0];
+		if (this._selectionSelectables.has(target)) return;
+		this._selectionSelectables.set(target, target);
+	}
+
+	_handleSelectionCheckboxUnsubscribe(e) {
+		e.stopPropagation();
+		const target = e.composedPath()[0];
+		this._selectionSelectables.delete(target);
+	}
+
+	_handleSelectionSelectAllChange(e) {
+		const checked = e.detail.checked;
+		this._selectionSelectables.forEach(selectable => selectable.checked = checked);
+		this._updateSelectionSubscribers();
+	}
+
+	_handleSelectionSubscriberSubscribe(e) {
+		e.stopPropagation();
+		const target = e.composedPath()[0];
+		if (this._selectionSubscribers.has(target)) return;
+		this._selectionSubscribers.set(target, target);
+		this._updateSelectionSubscribers();
+	}
+
+	_handleSelectionSubscriberUnsubscribe(e) {
+		e.stopPropagation();
+		const target = e.composedPath()[0];
+		this._selectionSubscribers.delete(target);
+	}
+
+	_updateSelectionSubscribers() {
+		if (!this._selectionSubscribers || this._selectionSubscribers.size === 0) return;
+
+		// debounce the updates for select-all case
+		if (this._updateSubscribersRequested) return;
+
+		this._updateSubscribersRequested = true;
+		setTimeout(() => {
+			const info = this.getSelectionInfo();
+			this._selectionSubscribers.forEach(subscriber => subscriber.selectionInfo = info);
```

Comment:
Curious... I think this is always going to trigger a re-render since the property will always change -- even if the underlying info hasn't. Maybe that's OK though since we're just talking about the summary and the select-all checkbox... which likely always change anyway.

Reply:
That's a good point, but yeah unless something's gone wrong, I think the info would always be different.  I could add a custom setter to the `SelectionSubscriberMixin` to not re-render in such a case.